### PR TITLE
Validation subsystem

### PR DESCRIPTION
**Pull request by rjplevin**
- Fixed typo in acid_gas_removal.py and implement required inputs/outputs for validation
- Implement required inputs/outputs in acid_gas_removal.py and bitumen_mining.py
- Documented how validation works (in readthedocs)
- Added support for tuples of required input/output stream contents, in which case at least one must exist
- Ran test suite and traced through new code to ensure it's working as intended

### CODE REVIEW

File: opgee/processes/acid_gas_removal.py [hunk 1]
```diff
@@ -61,6 +62,21 @@ class AcidGasRemoval(Process):
...
+            # TODO: If the process name were avoided, we could have just one output stream
+            #  with, say, "deacidified gas". Should describe the contents, not the destination.
+            # One of these must exist.
+            ("gas for demethanizer",  # TODO: avoid process names in contents
+             "gas for gas partition") # TODO: avoid process names in contents
+        ]
+
+        # Optional streams include:
+        # "gas for CO2 compressor" (output)
+
```
Comment (mbarlow-rmi) [inline opgee/processes/acid_gas_removal.py:79]: Regarding your comments, I agree that uniquely identifying the specific names seems brittle. My initial thought is that the declarations would lay out the number of streams and their contents, maybe with an optional identifier. What if we had something like:
```python
Phase = Literal['gas', 'solid', 'liquid']
class ProcessIOConstraint(NamedTuple):
    gas: int|List[Phase]
    oil: int
...
self._required_inputs: ProcessIOConstraints = ProcessIOConstraints(gas: 1)
```
There are likely ways of indicating "1 or more", "1 and only 1", or ">1,<=3", but it seems like the count and some kind of contents info would be necessary without using dynamic input values.

Comment (rjplevin) [reply opgee/processes/acid_gas_removal.py:79]: I'll have to think about this further, but my first impression is that this proposal is more complicated than necessary, and fails to address the original intent of the content strings.

The point of the stream "content" string is to allow processes to identify expected streams so they know which ones to read from and write to as required. (Before this, Processes would traverse the Stream to identify the Process on the other end.)

Anyway, I think a simple string accomplishes the original goal with minimal fuss. Having the validation code verify that the expected contents are present strikes me as the right level of validation.

Although the goal of avoiding dependencies on specific network structure in the Process subclass code was subverted by the inclusion of Process names in content strings, we can fix that pretty easily by renaming these.

Comment (mbarlow-rmi) [reply opgee/processes/acid_gas_removal.py:79]: It's already merged, but the main concern I have is that those names are set in the XML. Choosing a different name for a stream would then fail validation while potentially being a valid field. l think we'd want to either decouple the streams/processes from the input or remove the stream naming from the input.

**Comment by rjplevin** [comment]
Well, Process names already create unavoidable dependencies between the XML and the code. I don't see any way to avoid that, so the question is how to minimize it.

Note the we're not talking about *names* of Streams here: we're talking about general descriptions of their contents. This should be much less brittle than using specific Stream names, which by default includes the names of the Processes they join. (Streams only need to be named when there are multiple Streams connecting the same pair of Processes.)

The Stream's content-based tags are designed so that if a Process outputs, say, acidic gas to a Stream, it needs to go next to a process that consume acidic gas. That the first Process produces this type of output should be inherent to that process, so we're not creating a new dependency by labeling this. But there's no reason to depend on a specific consuming process, since it's conceivable that a new or client-specific process will be introduced that handles this. Any Process that can consume "acidic gas" should be able to take that Stream as an input. Indeed, it probably expects or requires acidic gas as an input.

Using a new "acid gas" consuming process requires that:
1. The XML be changed to identify this new Process by name 
2. The python module defining the class is found in the value of config variable ``OPGEE.ClassPath``
3. The XML is modified so the "acid gas" Stream connects to producer process to the consumer process.

So what have we achieved?
1. The producing process *code* is not affected by being connected to a novel process--it doesn't reference a specific process at the other end of any of its streams.
2. The contents of the stream from the producing process does not become a misnomer by referencing a (now defunct) consuming process, which would propagate confusion or require that the producing process (and XML) be updated to avoid this misnomer.

So, no, we definitely haven't avoided all XML/code dependencies, but we've avoided one whole class of them.

**Pull request merged**